src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1106 @@
+                Return the number 0, which index the unit of this algebra.
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(QQ).S()
+                    sage: A.one_basis()
+                    0
+                    sage: A.one()
+                    S[0]
+                """
+                return self.basis().keys()(0)
+
+            def degree_on_basis(self, m):
+                r"""
+                Return the degree of the basis element indexed by ``m``.
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(QQ).S()
+                    sage: A.degree_on_basis(4)
+                    4
+                """
+                return ZZ(m)
+
+            def from_polynomial(self, p):
+                """
+                Convert a polynomial into the ring of integer-valued polynomials.
+
+                This raises a ``ValueError`` if this is not possible.
+
+                INPUT:
+
+                - ``p`` -- a polynomial in one variable
+
+                EXAMPLES::
+
+                    sage: A = IntegerValuedPolynomialRing(ZZ).S()
+                    sage: S = A.basis()
+                    sage: S[5].polynomial()
+                    1/120*x^5 + 1/8*x^4 + 17/24*x^3 + 15/8*x^2 + 137/60*x + 1
+                    sage: A.from_polynomial(_)
+                    S[5]
+                    sage: x = polygen(QQ, 'x')
+                    sage: A.from_polynomial(x)
+                    -S[0] + S[1]
+
+                    sage: A = IntegerValuedPolynomialRing(ZZ).B()
+                    sage: B = A.basis()
+                    sage: B[5].polynomial()
+                    1/120*x^5 - 1/12*x^4 + 7/24*x^3 - 5/12*x^2 + 1/5*x
+                    sage: A.from_polynomial(_)
+                    B[5]
+                    sage: x = polygen(QQ, 'x')
+                    sage: A.from_polynomial(x)
+                    B[1]
+
+                TESTS::
+
+                    sage: x = polygen(QQ,'x')
+                    sage: A.from_polynomial(x+1/3)
+                    Traceback (most recent call last):
+                    ...
+                    ValueError: not a polynomial with integer values: 1/3
+                """
+                B = self.basis()
+                poly = self._poly
+                remain = p
+                result = self.zero()
+                while remain:
+                    N = remain.degree()
+                    top_coeff = remain.leading_coefficient() * factorial(N)
+                    try:
+                        top_coeff = self.base_ring()(top_coeff)
+                    except TypeError as exc:
+                        msg = 'not a polynomial with integer'
+                        msg += f' values: {top_coeff}'
+                        raise ValueError(msg) from exc
+                    remain += -top_coeff * poly(N)
+                    result += top_coeff * B[N]
+                return result
+
+            def gen(self):
```

Comment:
It would be good for this to take an (mostly ignored with a default of `0`) argument for consistency with other algebras/rings/etc.